File: src/main/java/com/intgrah/algorithms/heap/FibonacciHeap.java
package com.intgrah.algorithms.heap;

import com.intgrah.algorithms.list.Deletable;
import com.intgrah.algorithms.list.DoublyLinkedList;

import java.util.Comparator;
import java.util.HashMap;

public class FibonacciHeap<K> extends DecreasableHeap<K> {

    private final DoublyLinkedList<Node> roots = new DoublyLinkedList<>();
    private Node min = null;

    public FibonacciHeap(Comparator<K> ord) { super(ord); }

    @Override
    public Decreasable pushRef(K k) {
        Node n = new Node(k);
        addRoot(n);
        return n;
    }

    private void addRoot(Node n) {
        n.loser = false;
        n.parent = null;
        n.ref = roots.pushBack(n);
        if (min == null || ord.compare(n.key, min.key) < 0)
            min = n;
    }

    @Override
    public K getMin() {
        assert !isEmpty();
        return min.key;
    }

    @Override
    public void deleteMin() {
        assert !isEmpty();
        min.ref.delete();
        roots.append(min.children);
        HashMap<Integer, Node> d = new HashMap<>();
        for (Node n : roots) {
            int deg = n.children.size();
            while (d.containsKey(deg))
                n = link(n, d.remove(deg));
            d.put(deg, n);
        }
        roots.clear();
        min = null;
        for (Node n : d.values())
            addRoot(n);
    }

    @Override
    public boolean isEmpty() { return min == null; }

    @Override
    public void clear() {
        min = null;
        roots.clear();
    }

    private Node link(Node m, Node n) {
        if (ord.compare(m.key, n.key) < 0) {
            n.parent = m;
            n.ref = m.children.pushBack(n);
            return m;
        } else {
            m.parent = n;
            m.ref = n.children.pushBack(m);
            return n;
        }
    }

    private class Node extends Decreasable {

        private final DoublyLinkedList<Node> children = new DoublyLinkedList<>();
        private boolean loser = false;
        private Node parent = null;
        private Deletable ref;

        private Node(K k) { super(k); }

        @Override
        public void decreaseKey(K k) {
            assert ord.compare(k, key) <= 0;
            key = k;
            if (parent == null) {
                if (ord.compare(k, min.key) < 0)
                    min = this;
                return;
            }
            Node n = this;
            while (n.parent != null) {
                Node p = n.parent;
                n.ref.delete();
                addRoot(n);
                if (!p.loser) {
                    if (p.parent != null)
                        p.loser = true;
                    return;
                }
                n = p;
            }
        }

    }

}

File: heap/fibonacci_heap.py
from __future__ import annotations
from list.doubly_linked_list import *


class FibonacciNode[K]:
    def __init__(self, key: K) -> None:
        self.key = key
        self.loser = False
        self.parent: FibonacciNode[K] | None = None
        self.children = DoublyLinkedList[FibonacciNode[K]]()
        self.node: DoubleNode[FibonacciNode[K]]
        
    def merge(self, n: FibonacciNode[K]) -> FibonacciNode[K]:
        if self.key < n.key:
            n.parent = self
            n.node = self.children.push_back(n)
            return self
        else:
            self.parent = n
            self.node = n.children.push_back(self)
            return n


class FibonacciHeap[K]:
    def __init__(self) -> None:
        self.min: FibonacciNode[K] | None = None
        self.roots = DoublyLinkedList[FibonacciNode[K]]()
    
    def add_root(self, n: FibonacciNode[K]) -> None:
        n.loser = False
        n.parent = None
        n.node = self.roots.push_back(n)
        if self.min is None or n.key < self.min.key:
            self.min = n

    def push(self, key: K) -> FibonacciNode[K]:
        self.add_root(n := FibonacciNode(key))
        return n

    def pop_min(self) -> K:
        if (min := self.min) is None:
            raise IndexError
        self.roots.delete(min.node)
        d: dict[int, FibonacciNode[K]] = {}
        for n in list(self.roots) + list(min.children):
            while (deg := len(n.children)) in d:
                n = n.merge(d[deg])
                del d[deg]
            d[len(n.children)] = n
        self.roots.clear()
        self.min = None
        for n in d.values():
            self.add_root(n)
        return min.key
    
    def decrease_key(self, n: FibonacciNode[K], key: K) -> None:
        assert key <= n.key
        n.key = key
        if n.parent is None:
            if n.key < self.min.key:
                self.min = n
            return
        while (p := n.parent) is not None:
            p.children.delete(n.node)
            self.add_root(n)
            if not p.loser:
                if p.parent is not None:
                    p.loser = True
                return
            n = p
            
    def __bool__(self) -> bool:
        return self.min is not None

fh = FibonacciHeap[int]()
n5 = fh.push(5)
n6 = fh.push(6)
n8 = fh.push(8)
n7 = fh.push(7)
n1 = fh.push(1)
n3 = fh.push(3)
n2 = fh.push(2)
n4 = fh.push(4)
assert fh.pop_min() == 1
fh.decrease_key(n5, 0)
assert fh.pop_min() == 0
assert fh.pop_min() == 2
assert fh.pop_min() == 3
fh.decrease_key(n7, 5)
assert fh.pop_min() == 4
assert fh.pop_min() == 5
assert fh.pop_min() == 6
assert fh.pop_min() == 8
